File: dynamic_programming/rectangle_cutting/rectangle_cutting.cpp
#include <iostream>
#include <vector>
using namespace std;
#define MAX_CUTS 10000001

long long rectangle_cutting(int width, int height) {
    vector<vector<long long> > dp(width + 1, vector<long long>(height + 1, MAX_CUTS));
    for (int k = 0; k < min(width, height) + 1; k++) {
        dp[k][k] = 0;
    }
    for (int i = 1; i < dp.size(); i++) {
        for (int j = 1; j < dp[0].size(); j++) {
            for (int k = 1; k < j; k++) {
                dp[i][j] = min(dp[i][j], dp[i][k] + dp[i][j - k] + 1);
            }
            for (int k = 1; k < i; k++) {
                dp[i][j] = min(dp[i][j], dp[k][j] + dp[i - k][j] + 1);
            }
        }
    }
    return dp[width][height];
}

vector<int> read_nums(int size) {
    vector<int> nums(size);
    for (int i = 0; i < size; i++) {
        int num;
        scanf("%d", &num);
        nums[i] = num;
    }
    return nums;
}

int read_num() {
    int num;
    scanf("%d", &num);
    return num;
}

int main() {
    cout << rectangle_cutting(read_num(), read_num()) << endl;
}

File: dynamic_programming/rectangle_cutting/rectangle_cutting.py
from sys import stdin
import math

# TODO TLE O(n ^ 3) 
def rectangle_cutting(h, w):
    dp = [[0 for _ in range(h + 1)] for _ in range(w + 1)]
    for i in range(1, w + 1):
        for j in range(1, h + 1):
            if i == j:
                dp[i][j] = 0
                continue
            dp[i][j] = math.inf
            for k in range(1, i):
                dp[i][j] = min(dp[i][j], dp[k][j] + dp[i - k][j] + 1)
            for k in range(1, j):
                dp[i][j] = min(dp[i][j], dp[i][k] + dp[i][j - k] + 1)
    return dp[w][h]

def read_num():
    return int(stdin.readline())

def read_nums():
    return [int(num) for num in stdin.readline().split()]

if __name__ == '__main__':
    height, width = read_nums()
    print(rectangle_cutting(height, width))